crates/cli/src/native_query/aggregation_expression.rs

```diff
@@ -74,21 +127,185 @@ fn infer_type_from_aggregation_expression_document(
+                operand,
+            )?;
+            C::Scalar(BsonScalarType::Bool)
+        }
+        "$allElementsTrue" => {
+            infer_type_from_aggregation_expression(
+                context,
+                desired_object_type_name,
+                Some(&C::ArrayOf(Box::new(C::Scalar(BsonScalarType::Bool)))),
+                operand,
+            )?;
+            C::Scalar(BsonScalarType::Bool)
+        }
+        "$anyElementTrue" => {
+            infer_type_from_aggregation_expression(
+                context,
+                desired_object_type_name,
+                Some(&C::ArrayOf(Box::new(C::Scalar(BsonScalarType::Bool)))),
+                operand,
+            )?;
+            C::Scalar(BsonScalarType::Bool)
         }
-        (op, _) => Err(Error::UnknownAggregationOperator(op.to_string()))?,
+        "$arrayElemAt" => {
+            let (array_ref, idx) = two_parameter_operand(operator, operand)?;
+            let array_type = infer_type_from_aggregation_expression(
+                context,
+                &format!("{desired_object_type_name}_arrayElemAt_array"),
+                type_hint.map(|t| C::ArrayOf(Box::new(t.clone()))).as_ref(),
+                array_ref,
+            )?;
+            infer_type_from_aggregation_expression(
+                context,
+                &format!("{desired_object_type_name}_arrayElemAt_idx"),
+                Some(&C::Scalar(BsonScalarType::Int)),
+                idx,
+            )?;
+            type_hint
+                .cloned()
+                .unwrap_or_else(|| C::ElementOf(Box::new(array_type)))
+                .make_nullable()
+        }
+        "$split" => {
+            infer_types_from_aggregation_expression_tuple(
+                context,
+                desired_object_type_name,
+                Some(&C::Scalar(BsonScalarType::String)),
+                operand,
+            )?;
+            C::ArrayOf(Box::new(C::Scalar(BsonScalarType::String)))
+        }
+        op => Err(Error::UnknownAggregationOperator(op.to_string()))?,
     };
     Ok(t)
 }
 
+fn two_parameter_operand(operator: &str, operand: Bson) -> Result<(Bson, Bson)> {
+    match operand {
+        Bson::Array(operands) => {
+            if operands.len() != 2 {
+                return Err(Error::Other(format!(
+                    "argument to {operator} must be a two-element array"
+                )));
+            }
+            let mut operands = operands.into_iter();
+            let a = operands.next().unwrap();
+            let b = operands.next().unwrap();
+            Ok((a, b))
+        }
+        other_bson => Err(Error::ExpectedArrayExpressionArgument {
+            actual_argument: other_bson,
+        })?,
+    }
+}
+
+fn homogeneous_binary_operator_operand_type(
+    context: &mut PipelineTypeContext<'_>,
+    desired_object_type_name: &str,
+    operand_type_hint: Option<TypeConstraint>,
+    operator: &str,
+    operand: Bson,
+) -> Result<TypeConstraint> {
```

Comment:
So, broadly, forgive my massive handwaving, do I understand correctly that `TypeConstraint` contains either a concrete type or information we have about an unknown type (ie, that it's `Numeric`)?

And that we use subtyping to move constraints towards types when we combine them?

Reply:
Yes, that's pretty much it with the addendum that a `TypeConstraint` may reference a type variable. We have a type variable for the query result type, and another type variable for each parameter, and each variable is associated with a set of `TypeConstraint`s. Since `TypeConstraint` can reference variables that's a way for variables to indirectly reference each other.

Subtyping is part of it. I'm considering the query result variable to be covariant so multiple constraints converge on the supertype. Parameters are contravariant so those constraints should converge on the subtype, or lead to a mismatch error. But constraints can also relate variables in ways that are not related to subtyping. For example we've got `FieldOf` which is a way to state that the type of a variable is the same as the type of a given field of a given object type.